File: ts/src/action.ts
import {EventEmitter} from "node:events";
import {randomUUID} from "node:crypto";
import type {GrpcOptions, GrpcTransport} from "@protobuf-ts/grpc-transport";
import {ActionTransferRequest, type ActionNodeSubFlowValue, type ActionTransferResponse} from "@code0-tech/tucana/aquila";
import {constructValue, toAllowedValue, type PlainValue} from "@code0-tech/tucana/helpers";
import type {Value, Error as ProtoError} from "@code0-tech/tucana/shared";
import {RuntimeError} from "./types";
import type {DuplexStreamingCall} from "@protobuf-ts/runtime-rpc";
import type {FunctionClass} from "./models/function.model";
import type {RuntimeFunctionClass} from "./models/runtime_function.model";
import {runtimeFunctionMap} from "./map/runtime_function.map";
import {functionMap} from "./map/function.map";
import type {DataTypeClass} from "./models/datatype.model";
import {dataTypeMap} from "./map/datatype.map";
import type {EventClass} from "./models/event.model";
import {eventMap} from "./map/event.map";
import type {RuntimeEventClass} from "./models/runtime_event.model";
import {runtimeEventMap} from "./map/runtime_event.map";
import type {ConfigurationDefinition, Translation} from "./types";
import {CodeZeroEvent, type CodeZeroEventMap} from "./events";
import {createConnection} from "./internal/connection";
import {buildModule} from "./internal/module-builder";
import {ConfigManager} from "./manager/config-manager";
import {FlowManager} from "./manager/FlowManager";
import {FunctionManager} from "./manager/FunctionManager";
import {RuntimeFunctionManager} from "./manager/RuntimeFunctionManager";
import {DataTypeManager} from "./manager/DataTypeManager";
import {EventManager} from "./manager/EventManager";
import {RuntimeEventManager} from "./manager/RuntimeEventManager";
import {actions} from "./actions";

// Global registry keyed via Symbol.for so the CLI finds actions even when the
// user's code loaded a different copy of this module (e.g. the CJS build).
const ACTION_REGISTRY = Symbol.for("hercules.actions");

export function isExportMode(): boolean {
    return process.env.HERCULES_EXPORT === "1";
}

export function registeredActions(): readonly Action[] {
    return (globalThis as Record<symbol, unknown>)[ACTION_REGISTRY] as Action[] ?? [];
}

export class Action extends EventEmitter<CodeZeroEventMap> {
    private _transport?: GrpcTransport;
    private _stream?: DuplexStreamingCall<ActionTransferRequest, ActionTransferResponse>;
    private readonly _actions = new Map(actions.map(a => [a.packetType, a.handle]));
    // Pending sub flow / flow execution requests awaiting a response, keyed by a
    // per-invocation identifier: sub flow executions use the correlation
    // identifier generated for each invocation (echoed back on the response),
    // flow executions use their execution identifier. Both are unique per call,
    // so each key maps to a single pending request.
    private readonly _pendingExecutions = new Map<string, {
        resolve: (value: PlainValue) => void;
        reject: (error: unknown) => void;
    }>();

    readonly configs = new ConfigManager();
    readonly flows = new FlowManager();
    readonly functions = new FunctionManager();
    readonly runtimeFunctions = new RuntimeFunctionManager();
    readonly dataTypes = new DataTypeManager();
    readonly events = new EventManager();
    readonly runtimeEvents = new RuntimeEventManager();

    constructor(
        private readonly _identifier: string,
        private readonly _version: string,
        private readonly _aquilaUrl: string | undefined,
        private readonly _author: string,
        private readonly _icon: string,
        private readonly _documentation: string,
        private readonly _name: Translation[],
        private readonly _configurationDefinitions: ConfigurationDefinition[] = [],
    ) {
        super();
        if (isExportMode()) {
            const registry = (globalThis as Record<symbol, unknown>)[ACTION_REGISTRY] as Action[] | undefined;
            if (registry) registry.push(this);
            else (globalThis as Record<symbol, unknown>)[ACTION_REGISTRY] = [this];
        }
    }

    get identifier() { return this._identifier; }
    get version() { return this._version; }
    get stream() { return this._stream; }

    registerFunction<T extends RuntimeFunctionClass>(klass: FunctionClass<T>) {
        const def = functionMap(klass);
        this.functions.set(def.runtimeName, def);
    }

    registerRuntimeFunction(klass: RuntimeFunctionClass) {
        const omitDefinition = Reflect.getMetadata('hercules:omit_function_definition', klass) || false;
        const def = runtimeFunctionMap(klass);
        this.runtimeFunctions.set(def.runtimeName, def);
        if (!omitDefinition) {
            this.functions.set(def.runtimeName, {
                ...def,
                runtimeDefinitionName: def.runtimeName,
                parameters: def.parameters?.map(p => ({...p, runtimeDefinitionName: p.runtimeName})) || [],
            });
        }
    }

    registerDataTypeClass(klass: DataTypeClass) {
        const def = dataTypeMap(klass);
        this.dataTypes.set(def.identifier, def);
    }

    registerEventClass(klass: EventClass) {
        const def = eventMap(klass);
        this.events.set(def.identifier, def);
    }

    registerRuntimeEventClass(klass: RuntimeEventClass) {
        const omitDefinition = Reflect.getMetadata('hercules:omit_event_definition', klass) || false;
        const def = runtimeEventMap(klass);
        this.runtimeEvents.set(def.identifier, def);
        if (!omitDefinition) {
            this.events.set(def.identifier, {
                ...def,
                runtimeIdentifier: def.identifier,
            });
        }
    }

    /**
     * Fire the flow(s) bound to an event, or a single flow by id.
     *
     * Given an event class, every registered flow whose {@link ActionFlow.type}
     * matches the event's identifier is executed; the results are resolved
     * together as an array. Given a flow id, only that flow is executed and its
     * result is returned.
     *
     * Flows are tracked from the ActionFlowUpdate messages Aquila streams (see
     * {@link Action.flows}). Each flow is executed via {@link Action.executeFlow}.
     */
    fire(eventClass: EventClass | RuntimeEventClass, payload?: PlainValue): Promise<PlainValue[]>;
    fire(flowId: bigint | number, payload?: PlainValue): Promise<PlainValue>;
    async fire(
        target: EventClass | RuntimeEventClass | bigint | number,
        payload?: PlainValue,
    ): Promise<PlainValue | PlainValue[]> {
        if (!this._stream) throw new Error("Not connected. Call connect() first.");

        // Single flow by id.
        if (typeof target === "bigint" || typeof target === "number") {
            return this.executeFlow(target, payload);
        }

        // All flows bound to the event's flow type.
        const flowType = Reflect.getMetadata("hercules:identifier", target) as string | undefined;
        if (!flowType) throw new Error(`${target.name} is missing an @Identifier decorator.`);
        const flows = this.flows.filter(flow => flow.type === flowType);
        return Promise.all(flows.map(flow => this.executeFlow(flow.flowId, payload ?? null)));
    }

    /**
     * Execute the sub flow referenced by a {@link ActionNodeSubFlowValue} parameter
     * with the given parameters and resolve with its result. May be called
     * repeatedly for the same sub flow (e.g. once per iteration).
     */
    async executeSubFlow(subFlow: ActionNodeSubFlowValue, ...params: PlainValue[]): Promise<PlainValue> {
        if (!this._stream) throw new Error("Not connected. Call connect() first.");
        const {executionIdentifier} = subFlow;
        // Scope this individual invocation with a fresh correlation identifier so
        // its response can be matched even when the same sub flow is executed
        // repeatedly and responses arrive out of order.
        const correlationIdentifier = randomUUID();
        const result = this._awaitExecutionResponse(correlationIdentifier);
        const request = ActionTransferRequest.create({
            data: {
                oneofKind: "subFlowExecution",
                subFlowExecution: {
                    executionIdentifier,
                    parameters: params.map(p => constructValue(p ?? null)),
                    correlationIdentifier,
                },
            },
        });
        await this._stream.requests.send(request);
        this.emit(CodeZeroEvent.streamMessageSent, request);
        return result;
    }

    /**
     * Execute one of the action's own flows by id and resolve with its result.
     */
    async executeFlow(flowId: string | bigint | number, payload?: PlainValue): Promise<PlainValue> {
        if (!this._stream) throw new Error("Not connected. Call connect() first.");
        const executionIdentifier = randomUUID();
        const result = this._awaitExecutionResponse(executionIdentifier);
        const request = ActionTransferRequest.create({
            data: {
                oneofKind: "flowExecution",
                flowExecution: {
                    executionIdentifier,
                    flowId: String(flowId),
                    payload: constructValue(payload ?? null),
                },
            },
        });
        await this._stream.requests.send(request);
        this.emit(CodeZeroEvent.streamMessageSent, request);
        return result;
    }

    private _awaitExecutionResponse(identifier: string): Promise<PlainValue> {
        return new Promise<PlainValue>((resolve, reject) => {
            this._pendingExecutions.set(identifier, {resolve, reject});
        });
    }

    /**
     * Resolve or reject a pending sub flow / flow execution request. Invoked by
     * the response handlers when Aquila reports an execution's outcome. The
     * identifier is a sub flow's correlation identifier or a flow execution's
     * execution identifier.
     */
    resolveExecutionResponse(
        identifier: string,
        result:
            | {oneofKind: "success"; success: Value}
            | {oneofKind: "failure"; failure: ProtoError}
            | {oneofKind: undefined},
    ): void {
        const pending = this._pendingExecutions.get(identifier);
        this._pendingExecutions.delete(identifier);
        if (!pending) {
            this.emit(CodeZeroEvent.error, new Error(
                `Received execution response for unknown execution identifier: ${identifier}`,
            ));
            return;
        }
        if (result.oneofKind === "success") {
            pending.resolve(toAllowedValue(result.success));
        } else if (result.oneofKind === "failure") {
            pending.reject(new RuntimeError(result.failure.code, result.failure.message));
        } else {
            pending.reject(new Error("Received execution response with no result"));
        }
    }

    async connect(authToken: string, aquilaUrl?: string, grpcOptions?: GrpcOptions) {
        if (isExportMode()) return;
        const url = aquilaUrl ?? this._aquilaUrl;
        if (!url) throw new Error("aquilaUrl must be provided in the constructor or connect()");

        try {
            const {transport, stream} = await createConnection(this.buildModule(), authToken, url, grpcOptions);
            this._transport = transport;
            this._stream = stream;
        } catch (err) {
            this.emit(CodeZeroEvent.error, err as Error);
            throw err;
        }

        this.emit(CodeZeroEvent.connected, this);
        await this._processStream();
    }

    private async _processStream() {
        for await (const message of this._stream!.responses) {
            this.emit(CodeZeroEvent.streamMessageReceived, message);
            const {data} = message;
            if (data.oneofKind === undefined) {
                this.emit(CodeZeroEvent.error, new Error("Received unknown message type from stream"));
                continue;
            }
            this._actions.get(data.oneofKind)?.(this, (data as Record<string, unknown>)[data.oneofKind]);
        }
    }

    buildModule() {
        return buildModule({
            identifier: this._identifier, version: this._version,
            author: this._author, icon: this._icon, documentation: this._documentation,
            name: this._name, configurationDefinitions: this._configurationDefinitions,
            dataTypes: this.dataTypes.values(),
            events: this.events.values(),
            runtimeEvents: this.runtimeEvents.values(),
            functions: this.functions.values(),
            runtimeFunctions: this.runtimeFunctions.values(),
        });
    }
}

File: py/hercules/action.py
"""The :class:`Action` — the SDK entry point (port of ``src/action.ts``)."""
from __future__ import annotations

import asyncio
import inspect
import os
import uuid
from dataclasses import replace
from typing import Callable, Dict, List, Optional

from hercules._metadata import get_metadata
from tucana.generated.aquila import action_pb2
from hercules._tucana.helpers import PlainValue, construct_value, to_allowed_value
from hercules.actions import actions as _packet_handlers
from hercules.events import CodeZeroEvent
from hercules.internal.connection import create_connection
from hercules.internal.module_builder import ModuleBuildData, build_module
from hercules.manager.config import ConfigManager
from hercules.manager.datatype import DataTypeManager
from hercules.manager.event import EventManager
from hercules.manager.flow import FlowManager
from hercules.manager.function import FunctionManager
from hercules.manager.runtime_event import RuntimeEventManager
from hercules.manager.runtime_function import RuntimeFunctionManager
from hercules.map.datatype import data_type_map
from hercules.map.event import event_map
from hercules.map.function import function_map
from hercules.map.runtime_event import runtime_event_map
from hercules.map.runtime_function import runtime_function_map
from hercules.models.event import EventModel
from hercules.models.function import FunctionProps
from hercules.types import ConfigurationDefinition, RuntimeError, Translation

# Registry of actions constructed while exporting, so the CLI can find them.
_REGISTERED_ACTIONS: List["Action"] = []


def is_export_mode() -> bool:
    return os.environ.get("HERCULES_EXPORT") == "1"


def registered_actions() -> List["Action"]:
    return list(_REGISTERED_ACTIONS)


class EventEmitter:
    """A minimal Node-style event emitter."""

    def __init__(self) -> None:
        self._listeners: Dict[str, List[Callable]] = {}

    def on(self, event, listener: Callable) -> "EventEmitter":
        self._listeners.setdefault(str(event), []).append(listener)
        return self

    def once(self, event, listener: Callable) -> "EventEmitter":
        def wrapper(*args):
            self.off(event, wrapper)
            return listener(*args)

        return self.on(event, wrapper)

    def off(self, event, listener: Callable) -> "EventEmitter":
        listeners = self._listeners.get(str(event))
        if listeners and listener in listeners:
            listeners.remove(listener)
        return self

    def emit(self, event, *args) -> bool:
        listeners = list(self._listeners.get(str(event), []))
        for listener in listeners:
            result = listener(*args)
            if inspect.isawaitable(result):
                asyncio.ensure_future(result)
        return bool(listeners)


class Action(EventEmitter):
    def __init__(
        self,
        identifier: str,
        version: str,
        aquila_url: Optional[str],
        author: str,
        icon: str,
        documentation: str,
        name: List[Translation],
        configuration_definitions: Optional[List[ConfigurationDefinition]] = None,
    ) -> None:
        super().__init__()
        self._identifier = identifier
        self._version = version
        self._aquila_url = aquila_url
        self._author = author
        self._icon = icon
        self._documentation = documentation
        self._name = name
        self._configuration_definitions = [
            c if isinstance(c, ConfigurationDefinition) else ConfigurationDefinition(**c)
            for c in (configuration_definitions or [])
        ]

        self._channel = None
        self._stream = None
        self._actions = {a.packet_type: a.handle for a in _packet_handlers}
        # Pending sub flow / flow execution requests awaiting a response, keyed by
        # a per-invocation identifier: sub flow executions use the correlation
        # identifier generated for each invocation (echoed back on the response),
        # flow executions use their execution identifier. Both are unique per call,
        # so each key maps to a single pending request.
        self._pending_executions: Dict[str, asyncio.Future] = {}

        self.configs = ConfigManager()
        self.flows = FlowManager()
        self.functions = FunctionManager()
        self.runtime_functions = RuntimeFunctionManager()
        self.data_types = DataTypeManager()
        self.events = EventManager()
        self.runtime_events = RuntimeEventManager()

        if is_export_mode():
            _REGISTERED_ACTIONS.append(self)

    @property
    def identifier(self) -> str:
        return self._identifier

    @property
    def version(self) -> str:
        return self._version

    @property
    def stream(self):
        return self._stream

    # --- registration ---------------------------------------------------------
    def register_function(self, klass: type) -> None:
        definition = function_map(klass)
        self.functions.set(definition.runtime_name, definition)

    def register_runtime_function(self, klass: type) -> None:
        omit_definition = get_metadata("hercules:omit_function_definition", klass) or False
        definition = runtime_function_map(klass)
        self.runtime_functions.set(definition.runtime_name, definition)
        if not omit_definition:
            self.functions.set(
                definition.runtime_name,
                FunctionProps(
                    runtime_definition_name=definition.runtime_name,
                    runtime_name=definition.runtime_name,
                    signature=definition.signature,
                    throws_error=definition.throws_error,
                    name=definition.name,
                    description=definition.description,
                    documentation=definition.documentation,
                    deprecation_message=definition.deprecation_message,
                    display_message=definition.display_message,
                    alias=definition.alias,
                    display_icon=definition.display_icon,
                    design=definition.design,
                    parameters=[
                        replace(p, runtime_definition_name=p.runtime_name)
                        for p in (definition.parameters or [])
                    ],
                ),
            )

    def register_data_type_class(self, klass: type) -> None:
        definition = data_type_map(klass)
        self.data_types.set(definition.identifier, definition)

    def register_event_class(self, klass: type) -> None:
        definition = event_map(klass)
        self.events.set(definition.identifier, definition)

    def register_runtime_event_class(self, klass: type) -> None:
        omit_definition = get_metadata("hercules:omit_event_definition", klass) or False
        definition = runtime_event_map(klass)
        self.runtime_events.set(definition.identifier, definition)
        if not omit_definition:
            self.events.set(
                definition.identifier,
                EventModel(
                    identifier=definition.identifier,
                    signature=definition.signature,
                    settings=definition.settings,
                    editable=definition.editable,
                    name=definition.name,
                    description=definition.description,
                    documentation=definition.documentation,
                    display_message=definition.display_message,
                    alias=definition.alias,
                    display_icon=definition.display_icon,
                    runtime_identifier=definition.identifier,
                ),
            )

    # --- runtime dispatch -----------------------------------------------------
    async def send(self, request) -> None:
        if self._stream is None:
            raise RuntimeError("STREAM_NOT_CONNECTED", "Not connected. Call connect() first.")
        await self._stream.write(request)

    async def fire(self, target, payload: PlainValue = None):
        """Fire the flow(s) bound to an event, or a single flow by id.

        ``fire(event_class, payload)`` executes every registered flow whose
        ``type`` matches the event class' identifier, resolving the results
        together as a list. ``fire(flow_id, payload)`` executes just that flow and
        returns its result. Flows are tracked from the ActionFlowUpdate messages
        Aquila streams (see :attr:`flows`); each is executed via
        :meth:`execute_flow`.
        """
        if self._stream is None:
            raise Exception("Not connected. Call connect() first.")

        # Single flow by id.
        if isinstance(target, int) and not isinstance(target, bool):
            return await self.execute_flow(target, payload)

        # All flows bound to the event's flow type.
        flow_type = get_metadata("hercules:identifier", target)
        if not flow_type:
            raise Exception(f"{target.__name__} is missing an @Identifier decorator.")
        flows = self.flows.filter(lambda flow, _key: flow.type == flow_type)
        return await asyncio.gather(*(self.execute_flow(flow.flow_id, payload) for flow in flows))

    async def execute_sub_flow(self, sub_flow, *params: PlainValue) -> PlainValue:
        if self._stream is None:
            raise Exception("Not connected. Call connect() first.")
        execution_identifier = sub_flow.execution_identifier
        # Scope this individual invocation with a fresh correlation identifier so
        # its response can be matched even when the same sub flow is executed
        # repeatedly and responses arrive out of order.
        correlation_identifier = str(uuid.uuid4())
        result = self._await_execution_response(correlation_identifier)
        request = action_pb2.ActionTransferRequest(
            sub_flow_execution=action_pb2.ActionSubFlowExecutionRequest(
                execution_identifier=execution_identifier,
                parameters=[
                    construct_value(p if p is not None else None) for p in params
                ],
                correlation_identifier=correlation_identifier,
            )
        )
        await self.send(request)
        self.emit(CodeZeroEvent.stream_message_sent, request)
        return await result

    async def execute_flow(self, flow_id, payload: Optional[PlainValue] = None) -> PlainValue:
        if self._stream is None:
            raise Exception("Not connected. Call connect() first.")
        execution_identifier = str(uuid.uuid4())
        result = self._await_execution_response(execution_identifier)
        request = action_pb2.ActionTransferRequest(
            flow_execution=action_pb2.ActionFlowExecutionRequest(
                execution_identifier=execution_identifier,
                flow_id=str(flow_id),
                payload=construct_value(payload if payload is not None else None),
            )
        )
        await self.send(request)
        self.emit(CodeZeroEvent.stream_message_sent, request)
        return await result

    def _await_execution_response(self, identifier: str) -> asyncio.Future:
        future: asyncio.Future = asyncio.get_event_loop().create_future()
        self._pending_executions[identifier] = future
        return future

    def resolve_execution_response(self, identifier: str, response) -> None:
        pending = self._pending_executions.pop(identifier, None)
        if pending is None:
            self.emit(
                CodeZeroEvent.error,
                Exception(
                    f"Received execution response for unknown execution identifier: "
                    f"{identifier}"
                ),
            )
            return
        which = response.WhichOneof("result")
        if which == "success":
            pending.set_result(to_allowed_value(response.success))
        elif which == "failure":
            pending.set_exception(RuntimeError(response.failure.code, response.failure.message))
        else:
            pending.set_exception(Exception("Received execution response with no result"))

    async def connect(
        self,
        auth_token: str,
        aquila_url: Optional[str] = None,
        grpc_options: Optional[list] = None,
    ) -> None:
        if is_export_mode():
            return
        url = aquila_url or self._aquila_url
        if not url:
            raise Exception("aquila_url must be provided in the constructor or connect()")

        try:
            channel, stream = await create_connection(
                self.build_module(), auth_token, url, grpc_options
            )
            self._channel = channel
            self._stream = stream
        except Exception as err:  # noqa: BLE001
            self.emit(CodeZeroEvent.error, err)
            raise

        self.emit(CodeZeroEvent.connected, self)
        await self._process_stream()

    async def _process_stream(self) -> None:
        async for message in self._stream:
            self.emit(CodeZeroEvent.stream_message_received, message)
            which = message.WhichOneof("data")
            if which is None:
                self.emit(
                    CodeZeroEvent.error,
                    Exception("Received unknown message type from stream"),
                )
                continue
            handler = self._actions.get(which)
            if handler is not None:
                handler(self, getattr(message, which))

    def build_module(self):
        return build_module(
            ModuleBuildData(
                identifier=self._identifier,
                version=self._version,
                author=self._author,
                icon=self._icon,
                documentation=self._documentation,
                name=self._name,
                configuration_definitions=self._configuration_definitions,
                data_types=self.data_types.values(),
                events=self.events.values(),
                runtime_events=self.runtime_events.values(),
                functions=self.functions.values(),
                runtime_functions=self.runtime_functions.values(),
            )
        )
